### DISCUSSION TITLE
Upgrading from 1.5 -> 2

### DISCUSSION THREAD

**Discussion by matthew-moran** [elsa-workflows/elsa-core#1245] [Q&A] [answered]
I am trying to update some code to version 2.  I am having some difficulties and I was hoping I could get some help understanding the changes between the 2 versions.

In several of the activities OnExecuteAsync there is:
`context.CurrentScope.LastResult = Output.SetVariable("Metadata", fileMetadata);`
Im not sure how to change the code to get a similar behavior because LastResult no longer exists and Output's type has changed from Variables to OutputResult.

Also the workflows I'm upgrading have expressions that extend WorkflowExpression.  I can not find WorkflowExpression Is there someone that replaces it?  
Here is a example taken from a workflows build method 
`
.Then<FlattenPurchaseOrders>(x =>
                {
                    x.PurchaseOrders =  new CustomExpression<List<PurchaseOrder>>("@body('GetPurchaseOrders').PurchaseOrders");
                })`

`
public class CustomExpression<T> : WorkflowExpression<T>
    {
        public CustomExpression(string expression) : base(CustomScriptExpressionEvaluator.SyntaxName, expression)
        {
        }
    }
`

**Comment by sfmskywalker** [comment, marked as answer]
Hi @matthew-moran,

## 2.0

In 2.0, the "last result" concept has been replaced with "input", which is made available to the next activity based on the currently executing activity's output.

To provide output from an activity, you do so by using the `Output` method or one of the overloads of `Done` and `Outcome` - you will notice that these overloads accept an `output` parameter of type `object`.

Since output is of type `object`, you can set primitive as well as complex objects, including dictionaries.


## 2.1

In 2.1, you can still do the same as described for 2.0, but the recommended approach to providing activity output is by using **output properties**.

For example:

```csharp
[ActivityOutput] public MetaData Output { get; set; }
```

From your `OnExecute` method, you simply set a value to the `Output` property.

You can expose more than one output properties, but by convention, only the property named `Output` will automatically be available to the next activity as "input" (via the JS `input` variable).

If you need to access other output properties, you can do that too. This requires that the activity has a `Name` value set, allowing the following expression: `activities.SomeActivity1.MyOutput()` (notice that it is invoked as a function, not a property - this is due to a limitation in Jint) given the following output property declaration:

```csharp
[ActivityOutput] public MetaData MyOutput { get; set; }
```

In Liquid you would use the following expression:

```liquid
{{ Activities.SomeActivity1.MyOutput  }}
```

Contrary to the JS expression, here you access the property as a property, not as a function.

## Workflow Expressions
As of Elsa 2.0, you no longer declare activity input properties using the `WorkflowExpression<T>` type. Instead, it is now up to the engine to evaluate activity properties before executing the activity.

This means that from an activity's point of view, you are unaware of any expressions being executed - it simply receives the final values.

This does pose a bit of an inconvenience when using coded workflows like you seem to be doing when you actually want to use an expression that is something other than C#.

When coding workflows, one does typically something like this:

```csharp
.Then<FlattenPurchaseOrders>(activity => activity.Set(x => x.PurchaseOrders, context => GetPurchaseOrders(context)))
```

It would then be up to the `GetPurchaseOrders` method to provide a list of `PurchaseOrder` objects. This may or may not include custom script evaluation - that is up to you.
